piperider_cli/datasource.py

```diff
@@ -0,0 +1,311 @@
+
+    def _validate_required_fields(self):
+        reasons = []
+        # check required fields
+        for f in self.fields:
+            if f.name not in self.credential and f.optional is False:
+                reasons.append(f"{f.name} is required")
+
+        return reasons == [], reasons
+
+    @abstractmethod
+    def validate(self):
+        """
+        validate type name and required fields.
+
+        Returns True if everything is fine, False and reasons otherwise.
+
+        :return: bool, []
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def to_database_url(self):
+        """
+        build a database url for sqlalchemy create_engine method
+        :return:
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def verify_connector(self):
+        raise NotImplemented
+
+    def engine_args(self):
+        return dict()
+
+    def show_installation_information(self):
+        from rich.markup import escape
+        if self.verify_connector():
+            console = Console()
+            console.print(f'\n{escape(self.verify_connector())}\n')
+
+    def ask_credential(self):
+        """
+        ask for user filling all fields.
+        """
+
+        if sys.platform == "darwin":
+            # change readchar key backspace
+            readchar.key.BACKSPACE = '\x7F'
+
+        questions = []
+        for f in self.fields:
+            questions.append(f.question())
+
+        answers = inquirer.prompt(questions, raise_keyboard_interrupt=True)
+        if answers:
+            for f in self.fields:
+                self.credential[f.name] = answers[f.name].strip()
+        return self.credential
+
+    @staticmethod
+    def ask():
+        source_choices = [(k, v) for k, v in DATASOURCE_PROVIDERS.items()]
+
+        if sys.platform == "darwin":
+            # change readchar key backspace
+            readchar.key.BACKSPACE = '\x7F'
+
+        questions = [
+            inquirer.Text('project_name',
+                          message='What is your project name? (alphanumeric only)',
+                          validate=lambda ans, x: re.match(r'^[a-zA-Z0-9]+$', x) is not None),
+            inquirer.List('type',
+                          message='What data source would you like to connect to?',
+                          choices=source_choices,
+                          ),
+        ]
+        answers = inquirer.prompt(questions, raise_keyboard_interrupt=True)
+        name = answers['project_name'].strip()
+        cls = answers['type']
+        return cls, name
+
+
+class PostgreSQLDataSource(DataSource):
+    def __init__(self, name, **kwargs):
+        super().__init__(name, 'postgres', **kwargs)
+        self.fields = [
+            TextField('host', description='Host URL'),
+            NumberField('port', default=5432, description='Port'),
+            TextField('user', description='Username'),
+            PasswordField('password', description='Password (hidden)'),
```

Comment:
We can remove ` (hidden)` since user input is masked with asterisk

Reply:
fixed!